### PR TITLE
perf(graphql): add configurable tag selectivity sorting PE-7054

### PR DESCRIPTION
**Pull request by djwhitt**
This change adds support for configuring tag join order via a TAG_SELECTIVITY environment variable containing JSON mapping tag names to selectivity weights. This is used when constructing SQLite tag join SQL to determine the order in which to perform the joins (most selective first). It also alters the table used to order results. If no tag is present with a selectivity weight of >= 0 and there is an owner or target address specified on the query the transactions or data item tables are used for sorting rather than the tags table. This prevents a lot of index scaning when low selectivity tags are combined with high a high selectivity owner/target address filter.

Timing of ArDrive `App-Name` query with an owner filter using old and new logic:

```
Old:
Run Time: real 0.144 user 0.097366 sys 0.044813

New:
Run Time: real 0.002 user 0.000836 sys 0.000993
```

This query is not egregiously bad in the first place, but the improvement is still noticeable.

Testing with variations on this GraphQL query:

```graphql
query {
  transactions(
    owners: ["zPZe0p1Or5Kc0d7YhpT5kBC-JUPcDzUPJeMz2FdFiy4"]
    tags: [
      { name: "App-Name", values: ["ArDrive-App"] }
    ]
  ) {
    pageInfo {
      hasNextPage
    }
    edges {
      cursor
      node {
        id
      }
    }
  }
}
```

SQL with no owner filter:

```sql
SELECT *
FROM (
  SELECT DISTINCT
    st.height AS height,
    st.block_transaction_index AS block_transaction_index,
    x'00' AS data_item_id,
    0 AS indexed_at,
    id,
    last_tx AS anchor,
    signature,
    target,
    CAST(reward AS TEXT) AS reward,
    CAST(quantity AS TEXT) AS quantity,
    CAST(data_size AS TEXT) AS data_size,
    content_type,
    owner_address,
    public_modulus,
    sb.indep_hash AS block_indep_hash,
    sb.block_timestamp AS block_timestamp,
    sb.previous_block AS block_previous_block,
    '' AS parent_id
  FROM stable_transactions st
  INNER JOIN stable_blocks sb ON st.height = sb.height
  INNER JOIN wallets w ON st.owner_address = w.address
  INNER JOIN stable_transaction_tags AS "0_0"
    ON st.block_transaction_index = "0_0".block_transaction_index
    AND st.height = "0_0".height
  WHERE
    "0_0".tag_name_hash = x'5489BDC8235F51DCD7DE9368901685A679C5719F'
    AND "0_0".tag_value_hash IN (x'03900A76B78B6E591C4FA730280603B76E11784A')
  ORDER BY
    "0_0".height DESC NULLS FIRST,
    "0_0".block_transaction_index DESC NULLS FIRST,
    3 DESC,
    indexed_at DESC,
    5 DESC
  LIMIT 11
)
UNION
SELECT *
FROM (
  SELECT DISTINCT
    sdi.height AS height,
    sdi.block_transaction_index AS block_transaction_index,
    sdi.id AS data_item_id,
    sdi.indexed_at AS indexed_at,
    id,
    anchor,
    signature,
    target,
    '' AS reward,
    '' AS quantity,
    CAST(data_size AS TEXT) AS data_size,
    content_type,
    owner_address,
    public_modulus,
    sb.indep_hash AS block_indep_hash,
    sb.block_timestamp AS block_timestamp,
    sb.previous_block AS block_previous_block,
    sdi.parent_id
  FROM bundles.stable_data_items sdi
  INNER JOIN stable_blocks sb ON sdi.height = sb.height
  INNER JOIN bundles.wallets w ON sdi.owner_address = w.address
  INNER JOIN stable_data_item_tags AS "0_0"
    ON sdi.block_transaction_index = "0_0".block_transaction_index
    AND sdi.height = "0_0".height
    AND sdi.id = "0_0".data_item_id
  WHERE
    "0_0".tag_name_hash = x'5489BDC8235F51DCD7DE9368901685A679C5719F'
    AND "0_0".tag_value_hash IN (x'03900A76B78B6E591C4FA730280603B76E11784A')
  ORDER BY
    "0_0".height DESC NULLS FIRST,
    "0_0".block_transaction_index DESC NULLS FIRST,
    "0_0".data_item_id DESC,
    indexed_at DESC,
    5 DESC
  LIMIT 11
)
ORDER BY
  1 DESC,
  2 DESC,
  3 DESC
LIMIT 11;
```

SQL with an owner filter:

```sql
SELECT *
FROM (
  SELECT DISTINCT
    st.height AS height,
    st.block_transaction_index AS block_transaction_index,
    x'00' AS data_item_id,
    0 AS indexed_at,
    id,
    last_tx AS anchor,
    signature,
    target,
    CAST(reward AS TEXT) AS reward,
    CAST(quantity AS TEXT) AS quantity,
    CAST(data_size AS TEXT) AS data_size,
    content_type,
    owner_address,
    public_modulus,
    sb.indep_hash AS block_indep_hash,
    sb.block_timestamp AS block_timestamp,
    sb.previous_block AS block_previous_block,
    '' AS parent_id
  FROM stable_transactions st
  INNER JOIN stable_blocks sb ON st.height = sb.height
  INNER JOIN wallets w ON st.owner_address = w.address
  INNER JOIN stable_transaction_tags AS "0_0"
    ON st.block_transaction_index = "0_0".block_transaction_index
    AND st.height = "0_0".height
  WHERE
    st.owner_address IN (x'CCF65ED29D4EAF929CD1DED88694F99010BE2543DC0F350F25E333D857458B2E')
    AND "0_0".tag_name_hash = x'5489BDC8235F51DCD7DE9368901685A679C5719F'
    AND "0_0".tag_value_hash IN (x'03900A76B78B6E591C4FA730280603B76E11784A')
  ORDER BY
    st.height DESC NULLS FIRST,
    st.block_transaction_index DESC NULLS FIRST,
    3 DESC,
    indexed_at DESC,
    5 DESC
  LIMIT 11
)
UNION
SELECT *
FROM (
  SELECT DISTINCT
    sdi.height AS height,
    sdi.block_transaction_index AS block_transaction_index,
    sdi.id AS data_item_id,
    sdi.indexed_at AS indexed_at,
    id,
    anchor,
    signature,
    target,
    '' AS reward,
    '' AS quantity,
    CAST(data_size AS TEXT) AS data_size,
    content_type,
    owner_address,
    public_modulus,
    sb.indep_hash AS block_indep_hash,
    sb.block_timestamp AS block_timestamp,
    sb.previous_block AS block_previous_block,
    sdi.parent_id
  FROM bundles.stable_data_items sdi
  INNER JOIN stable_blocks sb ON sdi.height = sb.height
  INNER JOIN bundles.wallets w ON sdi.owner_address = w.address
  INNER JOIN stable_data_item_tags AS "0_0"
    ON sdi.block_transaction_index = "0_0".block_transaction_index
    AND sdi.height = "0_0".height
    AND sdi.id = "0_0".data_item_id
  WHERE
    sdi.owner_address IN (x'CCF65ED29D4EAF929CD1DED88694F99010BE2543DC0F350F25E333D857458B2E')
    AND "0_0".tag_name_hash = x'5489BDC8235F51DCD7DE9368901685A679C5719F'
    AND "0_0".tag_value_hash IN (x'03900A76B78B6E591C4FA730280603B76E11784A')
  ORDER BY
    sdi.height DESC NULLS FIRST,
    sdi.block_transaction_index DESC NULLS FIRST,
    3 DESC,
    indexed_at DESC,
    5 DESC
  LIMIT 11
)
ORDER BY
  1 DESC,
  2 DESC,
  3 DESC
LIMIT 11;
```

### CODE REVIEW

**Comment by djwhitt** [comment]
Testing with variations on this GraphQL query:

```
query {
  transactions(
    owners: ["zPZe0p1Or5Kc0d7YhpT5kBC-JUPcDzUPJeMz2FdFiy4"]
    tags: [
      { name: "App-Name", values: ["ArDrive-App"] }
    ]
  ) {
    pageInfo {
      hasNextPage
    }
    edges {
      cursor
      node {
        id
      }
    }
  }
}
```

**Comment by djwhitt** [comment]
SQL when no owner filter is present:

```sql
SELECT *
FROM (
  SELECT DISTINCT
    st.height AS height,
    st.block_transaction_index AS block_transaction_index,
    x'00' AS data_item_id,
    0 AS indexed_at,
    id,
    last_tx AS anchor,
    signature,
    target,
    CAST(reward AS TEXT) AS reward,
    CAST(quantity AS TEXT) AS quantity,
    CAST(data_size AS TEXT) AS data_size,
    content_type,
    owner_address,
    public_modulus,
    sb.indep_hash AS block_indep_hash,
    sb.block_timestamp AS block_timestamp,
    sb.previous_block AS block_previous_block,
    '' AS parent_id
  FROM stable_transactions st
  INNER JOIN stable_blocks sb ON st.height = sb.height
  INNER JOIN wallets w ON st.owner_address = w.address
  INNER JOIN stable_transaction_tags AS "0_0"
    ON st.block_transaction_index = "0_0".block_transaction_index
    AND st.height = "0_0".height
  WHERE
    "0_0".tag_name_hash = x'5489BDC8235F51DCD7DE9368901685A679C5719F'
    AND "0_0".tag_value_hash IN (x'03900A76B78B6E591C4FA730280603B76E11784A')
  ORDER BY
    "0_0".height DESC NULLS FIRST,
    "0_0".block_transaction_index DESC NULLS FIRST,
    3 DESC,
    indexed_at DESC,
    5 DESC
  LIMIT 11
)
UNION
SELECT *
FROM (
  SELECT DISTINCT
    sdi.height AS height,
    sdi.block_transaction_index AS block_transaction_index,
    sdi.id AS data_item_id,
    sdi.indexed_at AS indexed_at,
    id,
    anchor,
    signature,
    target,
    '' AS reward,
    '' AS quantity,
    CAST(data_size AS TEXT) AS data_size,
    content_type,
    owner_address,
    public_modulus,
    sb.indep_hash AS block_indep_hash,
    sb.block_timestamp AS block_timestamp,
    sb.previous_block AS block_previous_block,
    sdi.parent_id
  FROM bundles.stable_data_items sdi
  INNER JOIN stable_blocks sb ON sdi.height = sb.height
  INNER JOIN bundles.wallets w ON sdi.owner_address = w.address
  INNER JOIN stable_data_item_tags AS "0_0"
    ON sdi.block_transaction_index = "0_0".block_transaction_index
    AND sdi.height = "0_0".height
    AND sdi.id = "0_0".data_item_id
  WHERE
    "0_0".tag_name_hash = x'5489BDC8235F51DCD7DE9368901685A679C5719F'
    AND "0_0".tag_value_hash IN (x'03900A76B78B6E591C4FA730280603B76E11784A')
  ORDER BY
    "0_0".height DESC NULLS FIRST,
    "0_0".block_transaction_index DESC NULLS FIRST,
    "0_0".data_item_id DESC,
    indexed_at DESC,
    5 DESC
  LIMIT 11
)
ORDER BY
  1 DESC,
  2 DESC,
  3 DESC
LIMIT 11;
```

**Comment by djwhitt** [comment]
SQL with an owner filter:

```sql
SELECT *
FROM (
  SELECT DISTINCT
    st.height AS height,
    st.block_transaction_index AS block_transaction_index,
    x'00' AS data_item_id,
    0 AS indexed_at,
    id,
    last_tx AS anchor,
    signature,
    target,
    CAST(reward AS TEXT) AS reward,
    CAST(quantity AS TEXT) AS quantity,
    CAST(data_size AS TEXT) AS data_size,
    content_type,
    owner_address,
    public_modulus,
    sb.indep_hash AS block_indep_hash,
    sb.block_timestamp AS block_timestamp,
    sb.previous_block AS block_previous_block,
    '' AS parent_id
  FROM stable_transactions st
  INNER JOIN stable_blocks sb ON st.height = sb.height
  INNER JOIN wallets w ON st.owner_address = w.address
  INNER JOIN stable_transaction_tags AS "0_0"
    ON st.block_transaction_index = "0_0".block_transaction_index
    AND st.height = "0_0".height
  WHERE
    st.owner_address IN (x'CCF65ED29D4EAF929CD1DED88694F99010BE2543DC0F350F25E333D857458B2E')
    AND "0_0".tag_name_hash = x'5489BDC8235F51DCD7DE9368901685A679C5719F'
    AND "0_0".tag_value_hash IN (x'03900A76B78B6E591C4FA730280603B76E11784A')
  ORDER BY
    st.height DESC NULLS FIRST,
    st.block_transaction_index DESC NULLS FIRST,
    3 DESC,
    indexed_at DESC,
    5 DESC
  LIMIT 11
)
UNION
SELECT *
FROM (
  SELECT DISTINCT
    sdi.height AS height,
    sdi.block_transaction_index AS block_transaction_index,
    sdi.id AS data_item_id,
    sdi.indexed_at AS indexed_at,
    id,
    anchor,
    signature,
    target,
    '' AS reward,
    '' AS quantity,
    CAST(data_size AS TEXT) AS data_size,
    content_type,
    owner_address,
    public_modulus,
    sb.indep_hash AS block_indep_hash,
    sb.block_timestamp AS block_timestamp,
    sb.previous_block AS block_previous_block,
    sdi.parent_id
  FROM bundles.stable_data_items sdi
  INNER JOIN stable_blocks sb ON sdi.height = sb.height
  INNER JOIN bundles.wallets w ON sdi.owner_address = w.address
  INNER JOIN stable_data_item_tags AS "0_0"
    ON sdi.block_transaction_index = "0_0".block_transaction_index
    AND sdi.height = "0_0".height
    AND sdi.id = "0_0".data_item_id
  WHERE
    sdi.owner_address IN (x'CCF65ED29D4EAF929CD1DED88694F99010BE2543DC0F350F25E333D857458B2E')
    AND "0_0".tag_name_hash = x'5489BDC8235F51DCD7DE9368901685A679C5719F'
    AND "0_0".tag_value_hash IN (x'03900A76B78B6E591C4FA730280603B76E11784A')
  ORDER BY
    sdi.height DESC NULLS FIRST,
    sdi.block_transaction_index DESC NULLS FIRST,
    3 DESC,
    indexed_at DESC,
    5 DESC
  LIMIT 11
)
ORDER BY
  1 DESC,
  2 DESC,
  3 DESC
LIMIT 11;
```

**Comment by djwhitt** [comment]
SQL with an owner filter and Drive-Id filter after App-Name filter:
```sql
SELECT *
FROM (
  SELECT DISTINCT
    st.height AS height,
    st.block_transaction_index AS block_transaction_index,
    x'00' AS data_item_id,
    0 AS indexed_at,
    id,
    last_tx AS anchor,
    signature,
    target,
    CAST(reward AS TEXT) AS reward,
    CAST(quantity AS TEXT) AS quantity,
    CAST(data_size AS TEXT) AS data_size,
    content_type,
    owner_address,
    public_modulus,
    sb.indep_hash AS block_indep_hash,
    sb.block_timestamp AS block_timestamp,
    sb.previous_block AS block_previous_block,
    '' AS parent_id
  FROM stable_transactions st
  INNER JOIN stable_blocks sb ON st.height = sb.height
  INNER JOIN wallets w ON st.owner_address = w.address
  INNER JOIN stable_transaction_tags AS "0_0" ON st.block_transaction_index = "0_0".block_transaction_index
    AND st.height = "0_0".height
  CROSS JOIN stable_transaction_tags AS "1_1" INDEXED BY stable_transaction_tags_transaction_id_idx
  WHERE
    st.owner_address IN (X'ccf65ed29d4eaf929cd1ded88694f99010be2543dc0f350f25e333d857458b2e')
    AND "0_0".tag_name_hash = X'44b11a651a1c70673f9ba9bf0d6cda4a145e6b71'
    AND "0_0".tag_value_hash IN (X'36efff10b750de1092fa659a7bb98e781594befd')
    AND "1_1".transaction_id = "0_0".transaction_id
    AND "1_1".tag_name_hash = X'5489bdc8235f51dcd7de9368901685a679c5719f'
    AND "1_1".tag_value_hash IN (X'03900a76b78b6e591c4fa730280603b76e11784a')
  ORDER BY
    "0_0".height DESC NULLS FIRST,
    "0_0".block_transaction_index DESC NULLS FIRST,
    3 DESC,
    indexed_at DESC,
    5 DESC
  LIMIT 11
)
UNION
SELECT *
FROM (
  SELECT DISTINCT
    sdi.height AS height,
    sdi.block_transaction_index AS block_transaction_index,
    sdi.id AS data_item_id,
    sdi.indexed_at AS indexed_at,
    id,
    anchor,
    signature,
    target,
    '' AS reward,
    '' AS quantity,
    CAST(data_size AS TEXT) AS data_size,
    content_type,
    owner_address,
    public_modulus,
    sb.indep_hash AS block_indep_hash,
    sb.block_timestamp AS block_timestamp,
    sb.previous_block AS block_previous_block,
    sdi.parent_id
  FROM bundles.stable_data_items sdi
  INNER JOIN stable_blocks sb ON sdi.height = sb.height
  INNER JOIN bundles.wallets w ON sdi.owner_address = w.address
  INNER JOIN stable_data_item_tags AS "0_0" ON sdi.block_transaction_index = "0_0".block_transaction_index
    AND sdi.height = "0_0".height
    AND sdi.id = "0_0".data_item_id
  CROSS JOIN stable_data_item_tags AS "1_1" INDEXED BY stable_data_item_tags_data_item_id_idx
  WHERE
    sdi.owner_address IN (X'ccf65ed29d4eaf929cd1ded88694f99010be2543dc0f350f25e333d857458b2e')
    AND "0_0".tag_name_hash = X'44b11a651a1c70673f9ba9bf0d6cda4a145e6b71'
    AND "0_0".tag_value_hash IN (X'36efff10b750de1092fa659a7bb98e781594befd')
    AND "1_1".data_item_id = "0_0".data_item_id
    AND "1_1".tag_name_hash = X'5489bdc8235f51dcd7de9368901685a679c5719f'
    AND "1_1".tag_value_hash IN (X'03900a76b78b6e591c4fa730280603b76e11784a')
  ORDER BY
    "0_0".height DESC NULLS FIRST,
    "0_0".block_transaction_index DESC NULLS FIRST,
    "0_0".data_item_id DESC,
    indexed_at DESC,
    5 DESC
  LIMIT 11
)
ORDER BY
  1 DESC,
  2 DESC,
  3 DESC
LIMIT 11;
```

File: src/config.ts [hunk 1]
```diff
@@ -169,6 +169,26 @@ export const MAX_DATA_ITEM_QUEUE_SIZE = +env.varOrDefault(
...
+      Message: 20,
+      'Drive-Id': 10,
+      Process: 10,
+      Recipient: 10,
+      'App-Name': -10,
+      'Content-Type': -10,
+      'Data-Protocol': -10,
+    }),
+  ),
+) as Record<string, number>;
```
[the remark this answers is not in the document]
Comment (djwhitt) [reply src/config.ts:190]: Eh, reasonable suggestion, but this is documented elsewhere.

**Review by dtfiedler — APPROVED** [review]
very cool

**Pull request merged**